src/goto-instrument/contracts/dynamic-frames/dfcc_dsl_wrapper_program.cpp

```diff
@@ -0,0 +1,1141 @@
+
+  PRECONDITION(requires_write_set_symbol_opt);
+  PRECONDITION(addr_of_requires_write_set_symbol_opt);
+
+  // call write_set_create(
+  //   requires_write_set,
+  //   assigns_size = 0,
+  //   frees_size = 0,
+  //   replacement_mode = false,
+  //   assume_requires_ctx = contract_mode == check,
+  //   assert_requires_ctx = contract_mode != check,
+  //   assume_ensures_ctx = false,
+  //   assert_ensures_ctx = false,
+  // )
+  const auto write_set = requires_write_set_symbol_opt->symbol_expr();
+  preamble.add(goto_programt::make_decl(write_set));
+
+  const auto address_of_write_set =
+    addr_of_requires_write_set_symbol_opt->symbol_expr();
+  preamble.add(goto_programt::make_decl(address_of_write_set));
+  preamble.add(goto_programt::make_assignment(
+    address_of_write_set, address_of_exprt(write_set)));
+
+  auto function_symbol =
+    library.dfcc_fun_symbol.at(dfcc_funt::WRITE_SET_CREATE).symbol_expr();
+  code_function_callt call(function_symbol);
+  auto &arguments = call.arguments();
+
+  // write set
+  arguments.emplace_back(address_of_write_set);
+
+  // max assigns clause size
+  arguments.emplace_back(from_integer(0, size_type()));
+
+  // max frees clause size
+  arguments.emplace_back(from_integer(0, size_type()));
+
+  // replacement mode
+  arguments.emplace_back((exprt)false_exprt());
+
+  if(contract_mode == dfcc_contract_modet::CHECK)
+  {
+    // assume_requires_ctx
+    arguments.emplace_back((exprt)true_exprt());
+
+    // assert_requires_ctx
+    arguments.emplace_back((exprt)false_exprt());
+  }
+  else
+  {
+    // assume_requires_ctx
+    arguments.emplace_back((exprt)false_exprt());
+
+    // assert_requires_ctx mode
+    arguments.emplace_back((exprt)true_exprt());
+  }
+
+  // assume_ensures_ctx mode
+  arguments.emplace_back((exprt)false_exprt());
+
+  // assert_ensures_ctx mode
+  arguments.emplace_back((exprt)false_exprt());
+
+  preamble.add(goto_programt::make_function_call(call));
+
+  // check for absence of allocation/deallocation in requires clauses
+  // ```c
+  // DECL __check_no_alloc: bool;
+  // CALL __check_no_alloc = check_empty_alloc_dealloc(write_set);
+  // ASSERT __check_no_alloc;
+  // DEAD __check_no_alloc: bool;
+  // ```
+  auto check_var = get_fresh_aux_symbol(
+                     bool_typet(),
+                     id2string(wrapper_symbol.name),
+                     "__no_alloc_dealloc_in_requires",
+                     source_locationt{}, // TODO add property class and comment
+                     wrapper_symbol.mode,
+                     goto_model.symbol_table)
+                     .symbol_expr();
+
+  postamble.add(goto_programt::make_decl(check_var));
+
+  postamble.add(goto_programt::make_function_call(code_function_callt{
+    check_var,
+    library
+      .dfcc_fun_symbol
+        [dfcc_funt::WRITE_SET_CHECK_ALLOCATED_DEALLOCATED_IS_EMPTY]
+      .symbol_expr(),
+    {address_of_write_set}}));
+
+  // TODO add property class on assertion source_location
```

Comment:
Any reason this cannot be addressed right away?

Reply:
the main reason is I don't know how to generate a source location that makes sense for these instructions which are really ghost code and not directly derived from other source locations.

Reply:
I would go for the location of code that the property is about. Or, failing that, the location of the assigns clause. Of course there can be multiple assigns clauses. I suspect we should be augmenting each write-set member with a location.

Reply:
assigns clauses do not really exist in the AST, only their contents. even if they were, since there could be several clauses, and we collect them in a set, there's no single location we can refer to.